### PR TITLE
Recursion still NOT working for custom file

### PR DESCRIPTION
**Pull request by ribeaud**

### CODE REVIEW

**Comment by ribeaud** [comment]
OK, you will see, the changes are non-trivial. Tests are still green. Additionally, I made some manual tests: changing working directory vs. moving _.custom.env_ one level up vs. specifying a custom file with and without a directory.
Because I am not that expert in **Python**, I was NOT able to cover all the cases via tests (which will be great). Or this would just cost me too much time.
Have a look and try out. Cheers and have a nice evening.

File: environs.py [hunk 1]
```diff
@@ -222,28 +222,34 @@ def read_env(
...
             # Will be a directory
-            start = os.path.join(os.path.abspath(caller_dir))
+            head = os.path.abspath(caller_dir)
+            tail = ".env"
         else:
-            # Could be directory or a file
-            start = path
+            # Given 'path' could be a file or a directory, could exist or NOT
+            (root, _) = os.path.splitext(path)
+            if root == path:
```
Comment (sloria) [inline environs.py]: See `os.path.isdir`.

**Comment by sloria** [comment]
Can you write a test for the failing case(s)?

File: tests/test_environs.py [hunk 2]
```diff
@@ -231,20 +232,52 @@ def test_read_env_not_found_with_verbose_warns(self, env):
...
+
+    @pytest.mark.parametrize("recurse", [True, False])
+    def test_read_env_recurse(self, recurse, env):
         if "CUSTOM_STRING" in os.environ:
             os.environ.pop("CUSTOM_STRING")
         assert env("CUSTOM_STRING", "default") == "default"  # sanity check
-        env.read_env("tests/.custom.env", recurse=False)
-        assert env("CUSTOM_STRING") == "foo"
+        env.read_env(".custom.env", recurse=recurse)
+        if recurse:
```
Comment (sloria) [inline tests/test_environs.py]: Drive-by: might not be worth parametrizing this test with this conditional.

**Comment by sloria** [comment]
Let's clarify what the expected behavior is. This

If I'm understanding this PR correctly, the implemented behavior is:

| recurse | path | behavior                                                                                                                 |
| ------- | ---- | ------------------------------------------------------------------------------------------------------------------------ |
| False   | str  | If path has an extension, assume it's a file. If not, assume it's a directory and search for `.env` within the directory |
| False   | None | Search for `.env` in caller's directory.                                                                                 |
| True    | str  | Same as above, but recurse up from the initial path.                                                                     |
| True    | None | Search for `.env` in caller's directory, recurse upwards if not found                                                    |

I'm concerned about the cases where `path` is a string and doesn't exist. We can't reliably determine whether the path is (1) path to a file or (2) a path to a directory with a `.env`. For example, what is the desired behavior for

```python
# config/env does not exist
env.read_env("config/env", recurse=True)
```

Perhaps we should only allow filenames to be passed to `env.read_env`. I think this would be simpler, both from an implementation and from a user POV. What do you think, @ribeaud ?

**Comment by ribeaud** [comment]
Good evening @sloria. Sorry, a bit late to the party. I was sick. I will have a look at it tomorrow...

**Comment by ribeaud** [comment]
I noticed the following:`_walk_to_root` throws an exception if given path does NOT exist. And I made following assumption: a path with `.` in it points to a file. So `config/env` would point to a directory and NOT a file.
Nevertheless I agree with you that the method should be simple and easy to understand. Or we should document the method accordingly.

**Comment by sloria** [comment]
I think we should only accept file paths as the first argument, rather than trying to guess if it's a file or directory. That is the most intuitive and meets all use cases. Can you make that change?

**Comment by nhooey** [comment]
@ribeaud Any progress on this getting completed?

**Comment by timoklimmer** [comment]
@sloria @ribeaud I am also eagerly waiting on a fix. Would be great to get the fix very soon. Thanks!

**Comment by sloria** [comment]
I won't have time to look at this within the immediate future and it looks like @ribeaud is busy as well. @nhooey @timoklimmer If either of you would like to send a new PR, I'd gladly review and merge it. Let's go with my suggestion above: https://github.com/sloria/environs/pull/101#issuecomment-529143519

> Perhaps we should only allow filenames to be passed to env.read_env. I think this would be simpler, both from an implementation and from a user POV.

**Comment by timoklimmer** [comment]
I have created a couple of PRs now but they are all failing so far. Latest is #150. It passes almost all checks but then has issues with lint. I don't know how to fix that. It would be great if you could take a quick look. Thanks!
